libgadget/sfr_eff.c

```diff
@@ -78,28 +99,75 @@ void cooling_and_starformation(void)
         }
 
         if(shall_we_star_form) {
+            int newstar = -1;
             if(All.QuickLymanAlphaProbability > 0) {
-                double mass_of_star = find_star_mass(p_i);
-                make_particle_star(p_i, mass_of_star);
-                stars_converted++;
-                sum_mass_stars += mass_of_star;
+                /*New star is always the same particle as the parent for quicklya*/
+                newstar = p_i;
             } else {
-                int spawn = starformation(p_i, &localsfr, &sum_sm, &sum_mass_stars);
-                if(spawn == 1)
-                    stars_converted ++;
-                if(spawn == 2)
-                    stars_spawned ++;
+                newstar = starformation(p_i, &localsfr, &sum_sm);
+            }
+            /*Add this particle to the stellar conversion queue if necessary.*/
+            if(newstar >= 0) {
+                int tid = omp_get_thread_num();
+                thrqueuesfr[tid][nqthrsfr[tid]] = newstar;
+                thrqueueparent[tid][nqthrsfr[tid]] = p_i;
+                nqthrsfr[tid]++;
             }
         }
         else
             cooling_direct(p_i);
     }
 
+    /*Merge step for the queue.*/
+    if(NewStars) {
+        NumNewStar = gadget_compact_thread_arrays(NewStars, thrqueuesfr, nqthrsfr, All.NumThreads);
+        int NumNewParent = gadget_compact_thread_arrays(NewParents, thrqueueparent, nqthrsfr, All.NumThreads);
+        if(NumNewStar != NumNewParent)
+            endrun(3,"%d new stars, but %d new parents!\n",NumNewStar, NumNewParent);
+        /*Shrink star memory as we keep it for the wind model*/
+        NewStars = myrealloc(NewStars, sizeof(int) * NumNewStar);
```

Comment:
If we limit the queue length to NumPart, then there is no need to shrink it.

Reply:
It is still worthwhile (and it is only incrementing a pointer anyway) because there are usually much less than nactive stars and we may need memory for the Winds structure.